### PR TITLE
support for clustering in the redis DEL command

### PR DESCRIPTION
**Pull request by Geal**
When using redis in cluster mode, if we are trying to use any comounding commands like MGET or DEL, we must first split the list of keys per cluster hash, then query each redis instance according to its hash.
This was [already implemented for MGET](https://github.com/apollographql/router/commit/2c506fe128ff6acdeab42c38fbdc2049ea4f7ebb), this PR fixes it for the DEL command used in invalidation.

Our integration tests cannot use redis clusters yet, so this was checked with manual tests. I can help in setting it up to reproduce the issue and fix if someone else wants to try

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-589]: https://apollographql.atlassian.net/browse/ROUTER-589?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by garypen — APPROVED** [review]

File: apollo-router/src/cache/redis.rs [hunk 1]
```diff
@@ -562,16 +562,29 @@ impl RedisCacheStorage {
     }
 
     pub(crate) async fn delete<K: KeyType>(&self, keys: Vec<RedisKey<K>>) -> Option<u32> {
-        self.inner
-            .del(keys)
-            .await
-            .map_err(|e| {
-                if !e.is_not_found() {
+        let mut h: HashMap<u16, Vec<String>> = HashMap::new();
```
Comment (bnjjj) [inline apollo-router/src/cache/redis.rs:565]:
```suggestion
        let mut h: HashMap<u16, Vec<String>> = HashMap::with_capacity(keys.len());
```

Comment (bnjjj) [reply apollo-router/src/cache/redis.rs:565]: Does it make sense ?

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/cache/redis.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/cache/redis.rs:565]: yes

Comment (Geal) [reply apollo-router/src/cache/redis.rs:565]: kinda

Comment (Geal) [reply apollo-router/src/cache/redis.rs:565]: there will be less entries in the hashmap than the number of keys

Comment (bnjjj) [reply apollo-router/src/cache/redis.rs:565]: Feel free to decide whatever is better. It was just a suggestion :)

**Pull request merged**